### ISSUE TITLE
PageTemplate breaks on non-ASCII char in URL

### ISSUE THREAD

**Issue by icemac** [zopefoundation/Zope#382]
If the `id` of a PageTemplate created TTW contains a non-ASCII char (allowed since #181) it renders an error message in the ZMI telling that there is something wrong with the template code. But the error persists after removing the default template code.

Python 2 and 3 are affected.

**Comment by dataflake** [comment]
The error originates in https://github.com/malthe/chameleon/blob/af889be0247e24c5df5574979c36bdf9e2ed4ba7/src/chameleon/zpt/template.py#L313:

``` python
    def digest(self, body, names):
        hex = super(PageTemplate, self).digest(body, names)
        digest = md5(hex.encode('ascii'))
        <...>
```

The value for `hex` is computed in the super class `digest` method , which contains this snippet at https://github.com/malthe/chameleon/blob/af889be0247e24c5df5574979c36bdf9e2ed4ba7/src/chameleon/template.py#L257:

``` python
    def digest(self, body, names):
        class_name = type(self).__name__.encode('utf-8')
        sha = pkg_digest.copy()
        sha.update(body.encode('utf-8', 'ignore'))
        sha.update(class_name)
        digest = sha.hexdigest()

        if self.filename is not BaseTemplate.filename:
            digest = os.path.splitext(self.filename)[0] + '-' + digest    # XXX

        return digest
```

`self.filename` contains the ZODB path to the page template, which may contain non-ASCII characters now, ergo the forced encoding to ASCII will fail.

I have grepped all over the place but could not find where a page template that is not a filesystem page template gets its `filename` attribute set. It may be possible to quote that value so that we end up with straight ASCII for it.

**Comment by dataflake** [comment]
I have opened malthe/chameleon#284 for @malthe and hope he will take a look at it. This is a blocker for us.

**Comment by dataflake** [comment]
Malthe fixed this in malthe/chameleon@77f00e48a9f7fca14f9233a2e0777562cbdff50d. Thank you for the quick help @malthe! I'll keep this open until the next Chameleon release and will pin to it.

**Comment by icemac** [comment]
Using Chameleon release including the fix in 6805216326451a2e340d2c3ef4e087f916f5e36e.
